batchproducer/batchproducer.go

```diff
@@ -233,15 +232,22 @@ func (b *batchProducer) Stop() error {
 }
 
 func (b *batchProducer) setRunning(running bool) {
-	b.runningMu.Lock()
-	defer b.runningMu.Unlock()
-	b.running = running
+	var newValue int32
+	if running {
+		newValue = 1
+	} else {
+		newValue = 0
+	}
+
+	oldValue := b.running
+
+	for swapped := false; !swapped; {
```

Comment:
Okay, so if I got this right, `Start()` is the 

I'd rewrite `Start()` and `Stop()` like this:

```go
func (b *batchProducer) Start() error {
  if !atomic.CompareAndSwapInt32(&b.running, 0, 1) {
    return ErrAlreadyRunning
  }
  // At this point, b.isRunning() will return true so we can launch the go-routine
  go b.run()

  return nil
}
```

I'd keep stop as an immediate operation and maybe throw in a `StopAndWait()` function that actually waits for the buffer to drain completely. Kind of like what we have [here](https://github.com/timehop/goth/blob/master/worker/sleeper.go#L75-L100) — this may be overkill. A simple `Stop()` that blocks until the queue is drained might be enough.

```objc
func (b *batchProducer) Stop() error {
  if !atomic.CompareAndSwapInt32(&b.running, 1, 0) {
    return ErrAlreadyStopped
  }
  // at this point b.isRunning() will return `false`
  b.stop <- true

  // wait for queue drain?
  return ???
}
```

And I'd get rid of any flag setting inside of `run()`, as that should be controlled entirely by `Start()` and `Stop()` (unless something within the run loop can cause the whole thing to stop, which doesn't seem to be the case.)